File: src/prototype/web/tests/unit/judgment_window.test.ts
// Tests for JudgmentWindowFSM and the six technique predicates.
// References docs/design/state_machines_v1.md §8.

import { describe, expect, it } from "vitest";
import {
  INITIAL_JUDGMENT_WINDOW,
  TIME_SCALE,
  WINDOW_TIMING,
  crossCollarConditions,
  evaluateAllTechniques,
  flowerSweepConditions,
  hipBumpConditions,
  omoplataConditions,
  scissorSweepConditions,
  tickJudgmentWindow,
  triangleConditions,
  type JudgmentContext,
  type Technique,
} from "../../src/state/judgment_window.js";
import { initialActorState, type ActorState } from "../../src/state/game_state.js";
import type { HandFSM } from "../../src/state/hand_fsm.js";
import type { FootFSM } from "../../src/state/foot_fsm.js";
import type { GripZone } from "../../src/input/intent.js";

// Helpers to synthesise actor states for predicate testing without routing
// through the full FSM machinery.
function grippedHand(side: "L" | "R", zone: GripZone, atMs = 0): HandFSM {
  return Object.freeze({
    side,
    state: "GRIPPED" as const,
    target: zone,
    stateEnteredMs: atMs,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}
function idleHand(side: "L" | "R"): HandFSM {
  return Object.freeze({
    side,
    state: "IDLE" as const,
    target: null,
    stateEnteredMs: 0,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}
function foot(side: "L" | "R", state: FootFSM["state"]): FootFSM {
  return Object.freeze({ side, state, stateEnteredMs: 0 });
}

function actor(over: Partial<ActorState> = {}): ActorState {
  return Object.freeze({
    ...initialActorState(0),
    ...over,
  });
}

function ctx(over: Partial<JudgmentContext> = {}): JudgmentContext {
  return {
    bottom: actor(),
    top: actor(),
    bottomHipYaw: 0,
    bottomHipPush: 0,
    sustainedHipPushMs: 0,
    ...over,
  };
}

// --- Per-technique predicates ---------------------------------------------

describe("SCISSOR_SWEEP predicate (§8.2)", () => {
  it("fires with both feet LOCKED + SLEEVE gripped 0.6 + break ≥ 0.4", () => {
    const c = ctx({
      bottom: actor({
        leftFoot: foot("L", "LOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: grippedHand("L", "SLEEVE_R"),
        rightHand: idleHand("R"),
      }),
      top: actor({ postureBreak: { x: 0.5, y: 0 } }),
    });
    expect(scissorSweepConditions(c, 0.8, 0)).toBe(true);
  });

  it("fails if grip strength below 0.6", () => {
    const c = ctx({
      bottom: actor({
        leftFoot: foot("L", "LOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: grippedHand("L", "SLEEVE_R"),
      }),
      top: actor({ postureBreak: { x: 0.5, y: 0 } }),
    });
    expect(scissorSweepConditions(c, 0.4, 0)).toBe(false);
  });

  it("fails if a foot is UNLOCKED", () => {
    const c = ctx({
      bottom: actor({
        leftFoot: foot("L", "UNLOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: grippedHand("L", "SLEEVE_R"),
      }),
      top: actor({ postureBreak: { x: 0.5, y: 0 } }),
    });
    expect(scissorSweepConditions(c, 0.8, 0)).toBe(false);
  });

  it("fails if posture break magnitude < 0.4", () => {
    const c = ctx({
      bottom: actor({
        leftFoot: foot("L", "LOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: grippedHand("L", "SLEEVE_R"),
      }),
      top: actor({ postureBreak: { x: 0.2, y: 0 } }),
    });
    expect(scissorSweepConditions(c, 0.8, 0)).toBe(false);
  });
});

describe("FLOWER_SWEEP predicate (§8.2)", () => {
  it("needs both feet LOCKED + WRIST gripped + sagittal ≥ 0.5", () => {
    const c = ctx({
      bottom: actor({
        leftFoot: foot("L", "LOCKED"),
        rightFoot: foot("R", "LOCKED"),
        rightHand: grippedHand("R", "WRIST_L"),
      }),
      top: actor({ postureBreak: { x: 0, y: 0.6 } }),
    });
    expect(flowerSweepConditions(c, 0, 0.5)).toBe(true);
  });
});

describe("TRIANGLE predicate (§8.2)", () => {
  it("needs one foot UNLOCKED + arm_extracted + collar gripped", () => {
    const c = ctx({
      bottom: actor({
        leftFoot: foot("L", "UNLOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: grippedHand("L", "COLLAR_R"),
      }),
      top: actor({ armExtractedLeft: true }),
    });
    expect(triangleConditions(c)).toBe(true);
  });

  it("rejects when both feet are LOCKED", () => {
    const c = ctx({
      bottom: actor({
        leftFoot: foot("L", "LOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: grippedHand("L", "COLLAR_R"),
      }),
      top: actor({ armExtractedLeft: true }),
    });
    expect(triangleConditions(c)).toBe(false);
  });
});

describe("OMOPLATA predicate (§8.2)", () => {
  it("requires sleeve-side sign match + sagittal ≥ 0.6 + yaw ≥ π/3", () => {
    const c = ctx({
      bottom: actor({
        leftHand: grippedHand("L", "SLEEVE_R"),
      }),
      top: actor({ postureBreak: { x: -0.3, y: 0.7 } }), // lateral neg ↔ L sleeve sign
      bottomHipYaw: Math.PI / 3 + 0.1,
    });
    expect(omoplataConditions(c)).toBe(true);
  });

  it("fails if hip yaw below π/3", () => {
    const c = ctx({
      bottom: actor({ leftHand: grippedHand("L", "SLEEVE_R") }),
      top: actor({ postureBreak: { x: -0.3, y: 0.7 } }),
      bottomHipYaw: Math.PI / 4,
    });
    expect(omoplataConditions(c)).toBe(false);
  });

  it("fails if lateral break sign doesn't match the sleeve-hand side", () => {
    const c = ctx({
      bottom: actor({ leftHand: grippedHand("L", "SLEEVE_R") }),
      top: actor({ postureBreak: { x: 0.3, y: 0.7 } }), // wrong sign
      bottomHipYaw: Math.PI / 3 + 0.1,
    });
    expect(omoplataConditions(c)).toBe(false);
  });
});

describe("HIP_BUMP predicate (§8.2)", () => {
  it("needs sagittal ≥ 0.7 AND sustained 300ms push", () => {
    const c = ctx({
      top: actor({ postureBreak: { x: 0, y: 0.8 } }),
      sustainedHipPushMs: 350,
    });
    expect(hipBumpConditions(c)).toBe(true);
  });

  it("rejects short push durations", () => {
    const c = ctx({
      top: actor({ postureBreak: { x: 0, y: 0.8 } }),
      sustainedHipPushMs: 100,
    });
    expect(hipBumpConditions(c)).toBe(false);
  });
});

describe("CROSS_COLLAR predicate (§8.2)", () => {
  it("needs both COLLAR GRIPPED ≥ 0.7 and break ≥ 0.5", () => {
    const c = ctx({
      bottom: actor({
        leftHand: grippedHand("L", "COLLAR_L"),
        rightHand: grippedHand("R", "COLLAR_R"),
      }),
      top: actor({ postureBreak: { x: 0.3, y: 0.5 } }),
    });
    expect(crossCollarConditions(c, 0.8, 0.8)).toBe(true);
  });

  it("fails if one hand is not on a COLLAR zone", () => {
    const c = ctx({
      bottom: actor({
        leftHand: grippedHand("L", "COLLAR_L"),
        rightHand: grippedHand("R", "SLEEVE_L"),
      }),
      top: actor({ postureBreak: { x: 0.3, y: 0.5 } }),
    });
    expect(crossCollarConditions(c, 0.8, 0.8)).toBe(false);
  });
});

describe("evaluateAllTechniques", () => {
  it("reports every currently-satisfied technique", () => {
    const c = ctx({
      bottom: actor({
        leftFoot: foot("L", "LOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: grippedHand("L", "COLLAR_L"),
        rightHand: grippedHand("R", "COLLAR_R"),
      }),
      top: actor({ postureBreak: { x: 0.3, y: 0.5 } }),
    });
    const list = evaluateAllTechniques(c, 0.9, 0.9);
    expect(list).toContain("CROSS_COLLAR" satisfies Technique);
  });
});

// --- FSM lifecycle --------------------------------------------------------

describe("JudgmentWindowFSM lifecycle (§8.1)", () => {
  it("CLOSED → OPENING on any satisfied technique (outside cooldown)", () => {
    const { next, events } = tickJudgmentWindow(
      INITIAL_JUDGMENT_WINDOW,
      ["SCISSOR_SWEEP"],
      { nowMs: 0, confirmedTechnique: null, dismissRequested: false },
    );
    expect(next.state).toBe("OPENING");
    expect(events[0]?.kind).toBe("WINDOW_OPENING");
  });

  it("OPENING → OPEN after 200ms and holds time_scale at 0.3", () => {
    let w = INITIAL_JUDGMENT_WINDOW;
    const cand: Technique[] = ["SCISSOR_SWEEP"];
    const tick = (n: number) =>
      tickJudgmentWindow(w, cand, { nowMs: n, confirmedTechnique: null, dismissRequested: false });

    w = tick(0).next;               // CLOSED → OPENING
    const t1 = tick(100);
    expect(t1.next.state).toBe("OPENING");
    expect(t1.timeScale).toBeGreaterThan(TIME_SCALE.open);
    expect(t1.timeScale).toBeLessThan(TIME_SCALE.normal);

    w = tick(200).next;
    expect(w.state).toBe("OPEN");
  });

  it("OPEN → CLOSING on confirmed technique", () => {
    let w = INITIAL_JUDGMENT_WINDOW;
    const cand: Technique[] = ["SCISSOR_SWEEP"];
    w = tickJudgmentWindow(w, cand, { nowMs: 0, confirmedTechnique: null, dismissRequested: false }).next;
    w = tickJudgmentWindow(w, cand, { nowMs: 200, confirmedTechnique: null, dismissRequested: false }).next;
    expect(w.state).toBe("OPEN");

    const res = tickJudgmentWindow(
      w,
      cand,
      { nowMs: 300, confirmedTechnique: "SCISSOR_SWEEP", dismissRequested: false },
    );
    expect(res.next.state).toBe("CLOSING");
    expect(res.events.some((e) => e.kind === "TECHNIQUE_CONFIRMED")).toBe(true);
  });

  it("OPEN → CLOSING (DISRUPTED) when all candidates lose their conditions", () => {
    let w = INITIAL_JUDGMENT_WINDOW;
    const cand: Technique[] = ["SCISSOR_SWEEP"];
    w = tickJudgmentWindow(w, cand, { nowMs: 0, confirmedTechnique: null, dismissRequested: false }).next;
    w = tickJudgmentWindow(w, cand, { nowMs: 200, confirmedTechnique: null, dismissRequested: false }).next;

    const res = tickJudgmentWindow(
      w,
      [], // all conditions collapsed
      { nowMs: 300, confirmedTechnique: null, dismissRequested: false },
    );
    expect(res.next.state).toBe("CLOSING");
    expect(res.events.some((e) => e.kind === "WINDOW_CLOSING" && "reason" in e && e.reason === "DISRUPTED")).toBe(true);
  });

  it("OPEN → CLOSING (TIMED_OUT) after openMaxMs", () => {
    let w = INITIAL_JUDGMENT_WINDOW;
    const cand: Technique[] = ["SCISSOR_SWEEP"];
    w = tickJudgmentWindow(w, cand, { nowMs: 0, confirmedTechnique: null, dismissRequested: false }).next;
    w = tickJudgmentWindow(w, cand, { nowMs: 200, confirmedTechnique: null, dismissRequested: false }).next;

    const t = 200 + WINDOW_TIMING.openMaxMs;
    const res = tickJudgmentWindow(
      w,
      cand,
      { nowMs: t, confirmedTechnique: null, dismissRequested: false },
    );
    expect(res.next.state).toBe("CLOSING");
    expect(res.events.some((e) => e.kind === "WINDOW_CLOSING" && "reason" in e && e.reason === "TIMED_OUT")).toBe(true);
  });

  it("CLOSING → CLOSED and honours cooldown for 400ms", () => {
    let w = INITIAL_JUDGMENT_WINDOW;
    const cand: Technique[] = ["SCISSOR_SWEEP"];
    // OPEN → CLOSING on confirm.
    w = tickJudgmentWindow(w, cand, { nowMs: 0, confirmedTechnique: null, dismissRequested: false }).next;
    w = tickJudgmentWindow(w, cand, { nowMs: 200, confirmedTechnique: null, dismissRequested: false }).next;
    w = tickJudgmentWindow(
      w,
      cand,
      { nowMs: 300, confirmedTechnique: "SCISSOR_SWEEP", dismissRequested: false },
    ).next;
    expect(w.state).toBe("CLOSING");

    const closedAt = 300 + WINDOW_TIMING.closingMs;
    w = tickJudgmentWindow(w, cand, { nowMs: closedAt, confirmedTechnique: null, dismissRequested: false }).next;
    expect(w.state).toBe("CLOSED");
    expect(w.cooldownUntilMs).toBe(closedAt + WINDOW_TIMING.cooldownMs);

    // Cooldown: technique satisfied but OPENING should not fire.
    const midCooldown = tickJudgmentWindow(
      w,
      cand,
      { nowMs: closedAt + 100, confirmedTechnique: null, dismissRequested: false },
    );
    expect(midCooldown.next.state).toBe("CLOSED");

    // After cooldown, a new OPENING is permitted.
    const afterCooldown = tickJudgmentWindow(
      w,
      cand,
      { nowMs: closedAt + WINDOW_TIMING.cooldownMs + 1, confirmedTechnique: null, dismissRequested: false },
    );
    expect(afterCooldown.next.state).toBe("OPENING");
  });

  it("candidate set is frozen at OPENING entry (§8.3)", () => {
    let w = INITIAL_JUDGMENT_WINDOW;
    w = tickJudgmentWindow(w, ["SCISSOR_SWEEP"], { nowMs: 0, confirmedTechnique: null, dismissRequested: false }).next;
    expect(w.candidates).toEqual(["SCISSOR_SWEEP"]);

    // A newly-satisfied technique later should NOT be added.
    w = tickJudgmentWindow(
      w,
      ["SCISSOR_SWEEP", "HIP_BUMP"],
      { nowMs: 100, confirmedTechnique: null, dismissRequested: false },
    ).next;
    expect(w.candidates).toEqual(["SCISSOR_SWEEP"]);
  });
});

File: src/prototype/unity/Assets/BJJSimulator/Tests/EditMode/JudgmentWindowTest.cs
// NUnit EditMode mirror of src/prototype/web/tests/unit/judgment_window.test.ts.
// Each [Test] corresponds to one it(...) case from the Stage 1 Vitest suite.
//
// Run from Unity Editor: Window → General → Test Runner → EditMode.

using System.Collections.Generic;
using NUnit.Framework;

namespace BJJSimulator.Tests
{
    [TestFixture]
    public class JudgmentWindowTest
    {
        // --- helpers ----------------------------------------------------------

        static HandFSM GrippedHand(HandSide side, GripZone zone, long atMs = 0) => new HandFSM
        {
            Side = side, State = HandState.Gripped, Target = zone,
            StateEnteredMs = atMs, ReachDurationMs = 0,
            LastParriedZone = GripZone.None, LastParriedAtMs = BJJConst.SentinelTimeMs,
        };

        static HandFSM IdleHand(HandSide side) => new HandFSM
        {
            Side = side, State = HandState.Idle, Target = GripZone.None,
            StateEnteredMs = 0, ReachDurationMs = 0,
            LastParriedZone = GripZone.None, LastParriedAtMs = BJJConst.SentinelTimeMs,
        };

        static FootFSM Foot(FootSide side, FootState state) =>
            new FootFSM { Side = side, State = state, StateEnteredMs = 0 };

        static ActorState Actor(
            HandFSM? lh = null, HandFSM? rh = null,
            FootFSM? lf = null, FootFSM? rf = null,
            Vec2 pb = default,
            bool armExtL = false, bool armExtR = false)
        {
            var init = GameStateOps.InitialActorState();
            return new ActorState
            {
                LeftHand          = lh ?? init.LeftHand,
                RightHand         = rh ?? init.RightHand,
                LeftFoot          = lf ?? init.LeftFoot,
                RightFoot         = rf ?? init.RightFoot,
                PostureBreak      = pb,
                Stamina           = 1f,
                ArmExtractedLeft  = armExtL,
                ArmExtractedRight = armExtR,
            };
        }

        static JudgmentContext Ctx(
            ActorState? bottom = null, ActorState? top = null,
            float hipYaw = 0f, float hipPush = 0f, float sustainedPushMs = 0f)
        {
            var a = GameStateOps.InitialActorState();
            return new JudgmentContext
            {
                Bottom             = bottom ?? a,
                Top                = top    ?? a,
                BottomHipYaw       = hipYaw,
                BottomHipPush      = hipPush,
                SustainedHipPushMs = sustainedPushMs,
            };
        }

        static (JudgmentWindow Next, List<JudgmentTickEvent> Events) Tick(
            JudgmentWindow w, Technique[] cand, long nowMs,
            Technique? confirm = null, bool dismiss = false)
        {
            var events = new List<JudgmentTickEvent>();
            float ts;
            var next = JudgmentWindowOps.Tick(w, cand, new JudgmentTickInput
            {
                NowMs = nowMs, ConfirmedTechnique = confirm, DismissRequested = dismiss,
            }, events, out ts);
            return (next, events);
        }

        static readonly Technique[] ScissorOnly = { Technique.ScissorSweep };
        static readonly Technique[] None        = System.Array.Empty<Technique>();

        // --- per-technique predicates -----------------------------------------

        [Test]
        public void ScissorSweep_Fires_BothFeetLocked_SleeveGripped_BreakAbove04()
        {
            var c = Ctx(
                bottom: Actor(
                    lf: Foot(FootSide.L, FootState.Locked),
                    rf: Foot(FootSide.R, FootState.Locked),
                    lh: GrippedHand(HandSide.L, GripZone.SleeveR)),
                top: Actor(pb: new Vec2(0.5f, 0f)));
            Assert.IsTrue(JudgmentWindowOps.ScissorSweepConditions(c, 0.8f, 0f));
        }

        [Test]
        public void ScissorSweep_Fails_LowGripStrength()
        {
            var c = Ctx(
                bottom: Actor(
                    lf: Foot(FootSide.L, FootState.Locked),
                    rf: Foot(FootSide.R, FootState.Locked),
                    lh: GrippedHand(HandSide.L, GripZone.SleeveR)),
                top: Actor(pb: new Vec2(0.5f, 0f)));
            Assert.IsFalse(JudgmentWindowOps.ScissorSweepConditions(c, 0.4f, 0f));
        }

        [Test]
        public void ScissorSweep_Fails_FootUnlocked()
        {
            var c = Ctx(
                bottom: Actor(
                    lf: Foot(FootSide.L, FootState.Unlocked),
                    rf: Foot(FootSide.R, FootState.Locked),
                    lh: GrippedHand(HandSide.L, GripZone.SleeveR)),
                top: Actor(pb: new Vec2(0.5f, 0f)));
            Assert.IsFalse(JudgmentWindowOps.ScissorSweepConditions(c, 0.8f, 0f));
        }

        [Test]
        public void ScissorSweep_Fails_LowPostureBreak()
        {
            var c = Ctx(
                bottom: Actor(
                    lf: Foot(FootSide.L, FootState.Locked),
                    rf: Foot(FootSide.R, FootState.Locked),
                    lh: GrippedHand(HandSide.L, GripZone.SleeveR)),
                top: Actor(pb: new Vec2(0.2f, 0f)));
            Assert.IsFalse(JudgmentWindowOps.ScissorSweepConditions(c, 0.8f, 0f));
        }

        [Test]
        public void FlowerSweep_Fires_BothFeetLocked_WristGripped_SagittalAbove05()
        {
            var c = Ctx(
                bottom: Actor(
                    lf: Foot(FootSide.L, FootState.Locked),
                    rf: Foot(FootSide.R, FootState.Locked),
                    rh: GrippedHand(HandSide.R, GripZone.WristL)),
                top: Actor(pb: new Vec2(0f, 0.6f)));
            Assert.IsTrue(JudgmentWindowOps.FlowerSweepConditions(c, 0f, 0.5f));
        }

        [Test]
        public void Triangle_Fires_OneFootUnlocked_ArmExtracted_CollarGripped()
        {
            var c = Ctx(
                bottom: Actor(
                    lf: Foot(FootSide.L, FootState.Unlocked),
                    rf: Foot(FootSide.R, FootState.Locked),
                    lh: GrippedHand(HandSide.L, GripZone.CollarR)),
                top: Actor(armExtL: true));
            Assert.IsTrue(JudgmentWindowOps.TriangleConditions(c));
        }

        [Test]
        public void Triangle_Fails_BothFeetLocked()
        {
            var c = Ctx(
                bottom: Actor(
                    lf: Foot(FootSide.L, FootState.Locked),
                    rf: Foot(FootSide.R, FootState.Locked),
                    lh: GrippedHand(HandSide.L, GripZone.CollarR)),
                top: Actor(armExtL: true));
            Assert.IsFalse(JudgmentWindowOps.TriangleConditions(c));
        }

        [Test]
        public void Omoplata_Fires_SleeveHand_SagittalAbove06_CorrectSign_YawAbovePiOver3()
        {
            var c = Ctx(
                bottom: Actor(lh: GrippedHand(HandSide.L, GripZone.SleeveR)),
                top: Actor(pb: new Vec2(-0.3f, 0.7f)),
                hipYaw: (float)(System.Math.PI / 3.0) + 0.1f);
            Assert.IsTrue(JudgmentWindowOps.OmoplataConditions(c));
        }

        [Test]
        public void Omoplata_Fails_LowYaw()
        {
            var c = Ctx(
                bottom: Actor(lh: GrippedHand(HandSide.L, GripZone.SleeveR)),
                top: Actor(pb: new Vec2(-0.3f, 0.7f)),
                hipYaw: (float)(System.Math.PI / 4.0));
            Assert.IsFalse(JudgmentWindowOps.OmoplataConditions(c));
        }

        [Test]
        public void Omoplata_Fails_WrongLateralSign()
        {
            var c = Ctx(
                bottom: Actor(lh: GrippedHand(HandSide.L, GripZone.SleeveR)),
                top: Actor(pb: new Vec2(0.3f, 0.7f)), // wrong sign
                hipYaw: (float)(System.Math.PI / 3.0) + 0.1f);
            Assert.IsFalse(JudgmentWindowOps.OmoplataConditions(c));
        }

        [Test]
        public void HipBump_Fires_SagittalAbove07_Sustained300ms()
        {
            var c = Ctx(top: Actor(pb: new Vec2(0f, 0.8f)), sustainedPushMs: 350f);
            Assert.IsTrue(JudgmentWindowOps.HipBumpConditions(c));
        }

        [Test]
        public void HipBump_Fails_ShortPush()
        {
            var c = Ctx(top: Actor(pb: new Vec2(0f, 0.8f)), sustainedPushMs: 100f);
            Assert.IsFalse(JudgmentWindowOps.HipBumpConditions(c));
        }

        [Test]
        public void CrossCollar_Fires_BothCollarGripped_BreakAbove05()
        {
            var c = Ctx(
                bottom: Actor(
                    lh: GrippedHand(HandSide.L, GripZone.CollarL),
                    rh: GrippedHand(HandSide.R, GripZone.CollarR)),
                top: Actor(pb: new Vec2(0.3f, 0.5f)));
            Assert.IsTrue(JudgmentWindowOps.CrossCollarConditions(c, 0.8f, 0.8f));
        }

        [Test]
        public void CrossCollar_Fails_OneHandNotOnCollar()
        {
            var c = Ctx(
                bottom: Actor(
                    lh: GrippedHand(HandSide.L, GripZone.CollarL),
                    rh: GrippedHand(HandSide.R, GripZone.SleeveL)),
                top: Actor(pb: new Vec2(0.3f, 0.5f)));
            Assert.IsFalse(JudgmentWindowOps.CrossCollarConditions(c, 0.8f, 0.8f));
        }

        [Test]
        public void EvaluateAllTechniques_ReportsSatisfiedTechniques()
        {
            var c = Ctx(
                bottom: Actor(
                    lf: Foot(FootSide.L, FootState.Locked),
                    rf: Foot(FootSide.R, FootState.Locked),
                    lh: GrippedHand(HandSide.L, GripZone.CollarL),
                    rh: GrippedHand(HandSide.R, GripZone.CollarR)),
                top: Actor(pb: new Vec2(0.3f, 0.5f)));
            var list = JudgmentWindowOps.EvaluateAllTechniques(c, 0.9f, 0.9f);
            bool hasCrossCollar = false;
            foreach (var t in list) if (t == Technique.CrossCollar) { hasCrossCollar = true; break; }
            Assert.IsTrue(hasCrossCollar);
        }

        // --- FSM lifecycle ---------------------------------------------------

        [Test]
        public void Closed_To_Opening_OnSatisfiedTechnique_OutsideCooldown()
        {
            var (next, events) = Tick(JudgmentWindow.Initial, ScissorOnly, 0L);
            Assert.AreEqual(JudgmentWindowState.Opening, next.State);
            bool hasOpening = false;
            foreach (var e in events) if (e.Kind == JudgmentEventKind.WindowOpening) { hasOpening = true; break; }
            Assert.IsTrue(hasOpening);
        }

        [Test]
        public void Opening_To_Open_After200ms_TimeScaleAt03()
        {
            var w = JudgmentWindow.Initial;
            var events = new List<JudgmentTickEvent>();
            float ts;
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 0 }, events, out ts);
            events.Clear();
            JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 100 }, events, out ts);
            Assert.Greater(ts, JudgmentWindowTimeScale.Open);
            Assert.Less(ts,    JudgmentWindowTimeScale.Normal);
            events.Clear();
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 200 }, events, out ts);
            Assert.AreEqual(JudgmentWindowState.Open, w.State);
        }

        [Test]
        public void Open_To_Closing_OnConfirmedTechnique()
        {
            var w = JudgmentWindow.Initial;
            var events = new List<JudgmentTickEvent>();
            float ts;
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 0 }, events, out ts);
            events.Clear();
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 200 }, events, out ts);
            events.Clear();
            var res = new List<JudgmentTickEvent>();
            var next = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput
            {
                NowMs = 300, ConfirmedTechnique = Technique.ScissorSweep,
            }, res, out ts);
            Assert.AreEqual(JudgmentWindowState.Closing, next.State);
            bool hasConfirmed = false;
            foreach (var e in res) if (e.Kind == JudgmentEventKind.TechniqueConfirmed) { hasConfirmed = true; break; }
            Assert.IsTrue(hasConfirmed);
        }

        [Test]
        public void Open_To_Closing_Disrupted_WhenAllCandidatesCollapse()
        {
            var w = JudgmentWindow.Initial;
            var events = new List<JudgmentTickEvent>();
            float ts;
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 0   }, events, out ts);
            events.Clear();
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 200 }, events, out ts);
            events.Clear();
            var res  = new List<JudgmentTickEvent>();
            var next = JudgmentWindowOps.Tick(w, None, new JudgmentTickInput { NowMs = 300 }, res, out ts);
            Assert.AreEqual(JudgmentWindowState.Closing, next.State);
            bool hasDisrupted = false;
            foreach (var e in res)
                if (e.Kind == JudgmentEventKind.WindowClosing && e.CloseReason == WindowCloseReason.Disrupted)
                    { hasDisrupted = true; break; }
            Assert.IsTrue(hasDisrupted);
        }

        [Test]
        public void Open_To_Closing_TimedOut_AfterOpenMaxMs()
        {
            var w = JudgmentWindow.Initial;
            var events = new List<JudgmentTickEvent>();
            float ts;
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 0 },   events, out ts);
            events.Clear();
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 200 }, events, out ts);
            events.Clear();
            int t = 200 + JudgmentWindowTiming.Default.OpenMaxMs;
            var res  = new List<JudgmentTickEvent>();
            var next = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = t }, res, out ts);
            Assert.AreEqual(JudgmentWindowState.Closing, next.State);
            bool hasTimedOut = false;
            foreach (var e in res)
                if (e.Kind == JudgmentEventKind.WindowClosing && e.CloseReason == WindowCloseReason.TimedOut)
                    { hasTimedOut = true; break; }
            Assert.IsTrue(hasTimedOut);
        }

        [Test]
        public void Closing_To_Closed_HonoursCooldown_400ms()
        {
            var w = JudgmentWindow.Initial;
            var events = new List<JudgmentTickEvent>();
            float ts;
            // Drive to OPEN.
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 0 },   events, out ts);
            events.Clear();
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = 200 }, events, out ts);
            events.Clear();
            // Confirm → CLOSING.
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput
            {
                NowMs = 300, ConfirmedTechnique = Technique.ScissorSweep,
            }, events, out ts);
            events.Clear();
            Assert.AreEqual(JudgmentWindowState.Closing, w.State);

            // CLOSING → CLOSED.
            long closedAt = 300 + JudgmentWindowTiming.Default.ClosingMs;
            w = JudgmentWindowOps.Tick(w, ScissorOnly, new JudgmentTickInput { NowMs = closedAt }, events, out ts);
            events.Clear();
            Assert.AreEqual(JudgmentWindowState.Closed, w.State);
            Assert.AreEqual(closedAt + JudgmentWindowTiming.Default.CooldownMs, w.CooldownUntilMs);

            // Mid-cooldown: technique satisfied but should not re-OPEN.
            var (midNext, _) = Tick(w, ScissorOnly, closedAt + 100L);
            Assert.AreEqual(JudgmentWindowState.Closed, midNext.State);

            // After cooldown: should re-OPEN.
            var (afterNext, _) = Tick(w, ScissorOnly, closedAt + JudgmentWindowTiming.Default.CooldownMs + 1);
            Assert.AreEqual(JudgmentWindowState.Opening, afterNext.State);
        }

        [Test]
        public void CandidateSet_FrozenAtOpeningEntry()
        {
            var w = JudgmentWindow.Initial;
            var (w1, _) = Tick(w, ScissorOnly, 0L);
            Assert.AreEqual(1, w1.Candidates.Length);
            Assert.AreEqual(Technique.ScissorSweep, w1.Candidates[0]);

            // Second tick with additional candidate should NOT expand the list.
            var (w2, _) = Tick(w1, new[] { Technique.ScissorSweep, Technique.HipBump }, 100L);
            Assert.AreEqual(1, w2.Candidates.Length);
            Assert.AreEqual(Technique.ScissorSweep, w2.Candidates[0]);
        }
    }
}
